### ISSUE TITLE
No static IPv6 in LX zones

### ISSUE THREAD

**Issue by calmh** [joyent/smartos-live#431]
Using autoconfiguration works (with the appropriate filtering turned off), by starting `/native/usr/lib/inet/in.ndpd`; however manual configuration seems impossible.

```
root@dl2:~# /native/sbin/ifconfig 
lo0: flags=2001000849<UP,LOOPBACK,RUNNING,MULTICAST,IPv4,VIRTUAL> mtu 8232 index 1
        inet 127.0.0.1 netmask ff000000 
eth0: flags=1000843<UP,BROADCAST,RUNNING,MULTICAST,IPv4> mtu 1500 index 2
        inet 172.16.32.21 netmask ffffff00 broadcast 172.16.32.255
        ether 72:c4:cc:e5:b3:12 
lo0: flags=2002000849<UP,LOOPBACK,RUNNING,MULTICAST,IPv6,VIRTUAL> mtu 8252 index 1
        inet6 ::1/128 
eth0: flags=2000841<UP,RUNNING,MULTICAST,IPv6> mtu 1500 index 2
        inet6 fe80::70c4:ccff:fee5:b312/10 
        ether 72:c4:cc:e5:b3:12 
eth0:1: flags=2080841<UP,RUNNING,MULTICAST,ADDRCONF,IPv6> mtu 1500 index 2
        inet6 2001:470:deeb:32:70c4:ccff:fee5:b312/64 
root@dl2:~# ifconfig eth0 add 2001:470:deeb:32::21/64
SIOCSIFADDR: No such device or address
root@dl2:~# ip addr add 2001:470:deeb:32::21/64 dev eth0
Cannot find device "eth0"
root@dl2:~# /native/sbin/ifconfig eth0 inet6 addif 2001:470:deeb:32::21/64
ifconfig: 2001:470:deeb:32::21/64: bad address
root@dl2:~# /native/sbin/ipadm create-addr -t -T static -a 2001:470:deeb:32::21/64 eth0/v6s
ipadm: Error in setting local address: Operation failed
root@dl2:~# 
```

joyent_20150320T034246Z, lx-ubuntu-14.04 20150320

**Comment by pfmooney** [comment]
Using the built-in linux (ifconfig/ip) utilities for network configuration is not yet supported.
The native ifconfig will work, but it must be chrooted to access certain assets at their expected paths.
The /sbin/ifconfig-native script is pushed into the zone for that purpose.

Let me know if that doesn't work for you.

**Comment by calmh** [comment]
Ah! Yep, works, after a fashion. :)

```
root@unifi4:~# ifconfig-native net0 inet6 addif 2001:470:deeb:32::22/64
Created new logical interface net0:1
```

Yay!

```
root@unifi4:~# ping6 2001:470:deeb:32::1
connect: Network is unreachable
```

Hmmm...

```
root@unifi4:~# ifconfig-native 
lo0: flags=2001000849<UP,LOOPBACK,RUNNING,MULTICAST,IPv4,VIRTUAL> mtu 8232 index 1
        inet 127.0.0.1 netmask ff000000 
net0: flags=1000843<UP,BROADCAST,RUNNING,MULTICAST,IPv4> mtu 1500 index 2
        inet 172.16.32.22 netmask ffffff00 broadcast 172.16.32.255
        ether d2:2f:fe:b:bf:20 
lo0: flags=2002000849<UP,LOOPBACK,RUNNING,MULTICAST,IPv6,VIRTUAL> mtu 8252 index 1
        inet6 ::1/128 
net0: flags=2000841<UP,RUNNING,MULTICAST,IPv6> mtu 1500 index 2
        inet6 fe80::d02f:feff:fe0b:bf20/10 
        ether d2:2f:fe:b:bf:20 
net0:1: flags=2000840<RUNNING,MULTICAST,IPv6> mtu 1500 index 2
        inet6 2001:470:deeb:32::22/64 
```

Ah, it's not `UP`.

```
root@unifi4:~# ifconfig-native net0:1 up
ifconfig: setifflags: SIOCGLIFFLAGS: net0:1: no such interface
```

After some more fiddling it turned out that adding `up` to the same command as `addif` does indeed work.

```
root@unifi4:~# ping6 2001:470:deeb:32::1
WARNING: your kernel is veeery old. No problems.
PING 2001:470:deeb:32::1(2001:470:deeb:32::1) 56 data bytes
64 bytes from 2001:470:deeb:32::1: icmp_seq=1 time=3.11 ms
```

"veeery old", huh. :)

**Comment by rmustacc** [comment]
You'll may also need to start the native ndpd for additional IPv6 related functionality ala router advertisements, etc.

**Comment by calmh** [comment]
Yep. Noted in the first sentence of the issue. :) This was all about getting the static stuff running.

**Comment by sjorge** [comment]
@calmh have you tried:
```
ifconfig-native net0 inet6 addif 2001:470:deeb:32::22/64 up
```

Or better yet, since in.ndpd will also be running:
```
ifconfig-native net0 inet6 addif 2001:470:deeb:32::22/64 preferred up
```

This will make the static IP the prefered one to be used for outgoing traffic.

**Comment by calmh** [comment]
Indeed I have, as noted above. :)

I'm not sure if this issue should be left as is as a reminder to make this more intuitive (perhaps as part of the ever missing support for v6 provisioning in vmadm) or closed as it's possible to hack around with the proper incantations...

**Comment by melloc** [comment]
You can leave this open. I'm currently working on IPv6 support for SmartOS, and will be moving onto lx-brand support once I wrap up smartos zones. If you're interested, I can upload some SmartOS and SDC test images later, once I'm at a computer. :)

**Comment by melloc** [comment]
I've uploaded builds for [SmartOS](http://us-east.manta.joyent.com/cody.mello/public/builds/OS-2994/platform-OS-2994-usb.bz2) and [SDC](http://us-east.manta.joyent.com/cody.mello/public/builds/OS-2994/sdc-usb-OS-2994.tgz). vmadm(1m) has a section on nics.*.ips, which describes how to use multiple IP addresses and IPv6 SLAAC/DHCPv6. I'm still working on static gateways, but specifying IPv6 addresses should work without issue.

/cc @wiedi @bahamat who I believe expressed interested in testing these out.

**Comment by sjorge** [comment]
ipadm form /native also works, FYI 

~ sjorge

**Comment by danielpo** [comment]
Hello,

I have tried everything in this thread, but i always get "operation not supported" when trying to ping my default gw:
# ping6 2001:470:dd04:1::1
WARNING: your kernel is veeery old. No problems.
PING 2001:470:dd04:1::1(2001:470:dd04:1::1) 56 data bytes
ping: recvmsg: Operation not supported

It happens both with slaac and static ip, i have tried both "allow_ip_spoofing": true,
 and to specify the ips manually with allowed_ips.

Any ideas what i might have missed? Ipv6 is working in kvm and smartos zones.
I'm using joyent_20150806T063417Z and lx-ubuntu 14.04 20150601

my config:
"nics": [
    {
      "interface": "eth0",
      "mac": "82:b3:c7:3d:27:c9",
      "vlan_id": 20,
      "nic_tag": "admin",
      "gateway": "159.X.X.X",
      "ip": "159.X.X.X",
      "netmask": "255.255.255.0",
      "allow_ip_spoofing": true,

**Comment by wiedi** [comment]
@melloc just experimented a bit with your builds - very cool!

I like the idea of nics.*.ips - it solves this issue very elegantly.

The man-page paragraph regarding KVM was not clear to me at first until I read that this deprecates nics.*.ip. Maybe this could be reworded a bit (Maybe state that for KVM only one entry in the form of an IPv4 address or "dhcp" is supported?).

A few observations:

- when trying to create a IPv6 only instance with ` "ips": ["addrconf"]` the zone also configures an IPv4 address with DHCP.

- creating an IPv6 only instance with two static addresses `
"ips": ["2a01:138:a015:15::7001/64", "2a01:138:a015:15::7002/64"]` adds those correctly. It also configures additional all-zero addresses for IPv4 and IPv6: `inet 0.0.0.0 netmask 0` and `inet6 ::/0 `

- multiple static IPv4 addresses work as expected

- multiple mixed static v4 and v6 addresses also work but there is also the additional `inet6 ::/0 `

- the spoofing protection prevents using "addrconf" addresses that get configured with SLAAC (DHCPv6 works of course). This could be mentioned in the man page. 

all in all very nice already! If the source is available I might be able to include this in our own builds for further testing.

**Comment by sjorge** [comment]
I suspect 
  * the spoofing protection prevents using "addrconf" addresses that get configured with SLAAC (DHCPv6 works of course). This could be mentioned in the man page. 

May be related to the inet6 ::/0, addrconf behaves weirdly when it can't do all the stuff it needed.
While poking the gz for my PR I somehow ended up with it configuring inet ::/0 + a subnet that was my prefix without any routes a few times. :/

**Comment by wiedi** [comment]
`uts/common/io/mac/mac_protect.c` currently only implements learning of DHCPv6 configured addresses. I am unsure if SLAAC addresses are not supported for a security reason or because simply no one has done it yet.

**Comment by bahamat** [comment]
Fixing SLAAC is http://smartos.org/bugview/OS-4667.

**Comment by melloc** [comment]
@wiedi Thanks for trying out the images and for the feedback!

I'll touch up that paragraph and try to make it clearer. Regarding the other points:

> when trying to create a IPv6 only instance with "ips": ["addrconf"] the zone also configures an IPv4 address with DHCP.

I'll change this. It should be possible to have an IPv6 only instance.

> creating an IPv6 only instance with two static addresses "ips": ["2a01:138:a015:15::7001/64", "2a01:138:a015:15::7002/64"] adds those correctly. It also configures additional all-zero addresses for IPv4 and IPv6: inet 0.0.0.0 netmask 0 and inet6 ::/0

The all-zero IPv4 address is probably an interface trying to do DHCP. I've filed a bug for this, and hope to fix it this week:

https://smartos.org/bugview/OS-4683

> multiple mixed static v4 and v6 addresses also work but there is also the additional inet6 ::/0

Does the interface say ADDRCONF? I would see this during testing when I wasn't on a network with router advertisements. Is this the case here? @sjorge What about your network setup?

> the spoofing protection prevents using "addrconf" addresses that get configured with SLAAC (DHCPv6 works of course). This could be mentioned in the man page.

I'm going to move onto adding SLAAC detection once I finish up several other things.

I'll put up a Git patch along with some new images later today if you want to apply it to your local builds.

**Comment by wiedi** [comment]
> Does the interface say ADDRCONF?

while the all-zero ipv6 address indeed has the ADDRCONF flag the all-zero ipv4 does not show DHCP.
In my tests it did not matter if ipv6 router advertisements or ipv4 dhcp were actually offered.

**Comment by melloc** [comment]
Resolved by joyent/illumos-joyent@632f82b. As mentioned in #231, [OS-4667](https://smartos.org/bugview/OS-4667) and [OS-4683](https://smartos.org/bugview/OS-4683) may affect those using IP spoofing protection until they are resolved.
